File: dms613412/Programmars/Level4/호텔 방 배정.cpp

#include <string>
#include <vector>
#include <map>

using namespace std;

map <long long, long long> m;
long long find(long long x) {
	if (!m[x])return x;
	m[x] = find(m[x]);
	return find(m[x]);
}

vector<long long> solution(long long k, vector<long long> room_number) {

	vector<long long> answer;
	for (auto &x : room_number) {
		if (!m[x]) {
			answer.push_back(x);
			m[x] = find(x + 1);
		}
		else {
			long long tmp = find(m[x]);
			answer.push_back(tmp);
			m[tmp] = find(tmp + 1);
		}
	}

	return answer;
}

File: dms613412/Programmars/Level4/호텔방배정.py
import sys
sys.setrecursionlimit(100000) #파이썬3에는 기본 재귀 제한이 있고 이게 약 1000회

m = {}
def find(x):
    if not x in m.keys():
        return x
    m[x]=find(m[x])
    return m[x]

def solution(k, room_number):
    answer = []

    for x in room_number:
        if x in m.keys():
            tmp = find(m[x])
            answer.append(tmp)
            m[tmp]=find(tmp+1)            
        else:
            answer.append(x)
            m[x]=find(x+1)    
    
    return answer
